8319795: Static huge pages are not used for CodeCache

eastig opened the pull request:
This is a fix of a regression introduced by [JDK-8261894](https://bugs.openjdk.org/browse/JDK-8261894).
After JDK-8261894, `os::can_execute_large_page_memory()` returns `true` only if `UseTransparentHugePages` is true. As CodeCache uses `os::can_execute_large_page_memory()` when it detects a page size, CodeCache cannot use static huge pages (`UseTransparentHugePages` is `false`) anymore after the change.
Before JDK-8261894, `os::can_execute_large_page_memory()` returned `true`  when either `UseTransparentHugePages` or `UseHugeTLBFS` was true.

After JDK-8261894, `XX:+UseLargePages XX:-UseTransparentHugePages` means to use static huge pages: aka `UseHugeTLBFS` is `true`. If `UseLargePages` is not set to `true` via the option, it will be set to `true` if `UseTransparentHugePages` is `true`.

`os::can_execute_large_page_memory()` is modified to return `UseLargePages`. A regression gtest  is added.

Tested fastdebug and release builds:
- [x] tier1
- [x] gtest
- [x] test/hotspot/jtreg/gtest/LargePageGtests.java

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319795](https://bugs.openjdk.org/browse/JDK-8319795): Static huge pages are not used for CodeCache (**Bug** - P3)


### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [d1bdfdfd](https://git.openjdk.org/jdk/pull/16582/files/d1bdfdfdf2262f360387ca5adbc530776acb7421)
 * [Volker Simonis](https://openjdk.org/census#simonis) (@simonis - **Reviewer**) ⚠️ Review applies to [d1bdfdfd](https://git.openjdk.org/jdk/pull/16582/files/d1bdfdfdf2262f360387ca5adbc530776acb7421)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16582/head:pull/16582` \
`$ git checkout pull/16582`

Update a local copy of the PR: \
`$ git checkout pull/16582` \
`$ git pull https://git.openjdk.org/jdk.git pull/16582/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16582`

View PR using the GUI difftool: \
`$ git pr show -t 16582`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16582.diff">https://git.openjdk.org/jdk/pull/16582.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16582#issuecomment-1804048016)

shipilev reviewed (approved):
Looks like a pretty simple regression.
@tstuefe needs to take a look as well.

simonis reviewed (approved):
Good catch! Looks good to me.

tstuefe commented:
Taking a look, let me think for a bit..

tstuefe commented:
Good catch.

But the whole function makes no sense anymore. It was used because System V shared memory could not be set executable. We don't use that anymore, so any large page memory is executable. You could return true now (returning UseLargePages makes not that much sense).

But no other platform has restrictions like that. Windows has large pages, and they can be set executable. Now Linux is the same. BSD/MacOs and AIX don't use large pages. So I think we can just completely remove that function.

tstuefe commented:
We should have a regression test for this...

eastig commented:
@tstuefe Yes, removing the function makes sense to me. Without the function another regression test is needed because the one in this PR will be removed. I'll add the test.

tstuefe commented:
> @tstuefe Yes, removing the function makes sense to me. Without the function another regression test is needed because the one in this PR will be removed. I'll add the test.

Cool. BTW the test does not have to be complicated. 

Just start a VM with `java -Xmx64m -Xlog:pagesize* -XX:+UseLargePages -version` and scan the output.

If the output shows `UseLargePages=1`, the VM uses large pages, which means the test system has large pages. In that case, we also expect something like this:

`CodeHeap .. blabla .. page_size=2M`

eastig commented:
Hm...
I looked into `CodeCache::page_size`, its history and uses. The function smells. It has an interesting hack for the large page case:
```
  if (os::can_execute_large_page_memory()) {
    if (InitialCodeCacheSize < ReservedCodeCacheSize) {
      // Make sure that the page size allows for an incremental commit of the reserved space
      min_pages = MAX2(min_pages, (size_t)8);
    }
``` 

The uses are:
- 2 of `page_size(false, 8)`
- 1 of `page_size(false, 1)`
- 1 of `page_size(true, 1)`

This looks strange to me. I need to if everything is correct.

tstuefe commented:
> Hm... I looked into `CodeCache::page_size`, its history and uses. The function smells. It has an interesting hack for the large page case:
> 
> ```
>   if (os::can_execute_large_page_memory()) {
>     if (InitialCodeCacheSize < ReservedCodeCacheSize) {
>       // Make sure that the page size allows for an incremental commit of the reserved space
>       min_pages = MAX2(min_pages, (size_t)8);
>     }
> ```
> 
> The uses are:
> 
>     * 2 of `page_size(false, 8)`
> 
>     * 1 of `page_size(false, 1)`
> 
>     * 1 of `page_size(true, 1)`
> 
> 
> This looks strange to me. I need to if everything is correct.

I'm not the original author, but my guess is this tries to prevent that if the large page size is very large, e.g. 1G, it should use smaller page sizes, e.g. 2M or 4K, because otherwise a CodeCacheSize of 1G would allocate 1x1GB page, and that would be fully committed right from the start and increase memory footprint.

Seems like an odd optimization though. It feels a bit arbitrary since we don't do this for the java heap, for example. Also, it prevents users from ever using e.g. 1G pages for code cache if they really want that.

@TobiHartmann ?

As for the `if (os::can_execute_large_page_memory()) {`, I believe that was slightly wrong. It feels like it wants to guard the large-page case, but on Windows "can_execute_large_page_memory" is always true, regardless of UseLargePages. And now, on Linux, it is also always true.

I would just swap that with `if (UseLargePages)`.

TobiHartmann commented:
Right, I think that was the original intention and I agree that it seems inconsistent. We should clean that code up.

eastig commented:
@tstuefe, @TobiHartmann,
`InitialCodeCacheSize` is useless when static huge pages are used. When static huge pages are reserved, they are committed: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/memory/virtualspace.cpp#L248
```
  if (use_explicit_large_pages(page_size)) {
    // System can't commit large pages i.e. use transparent huge pages and
    // the caller requested large pages. To satisfy this request we use
    // explicit large pages and these have to be committed up front to ensure
    // no reservations are lost.
    do {
```

tstuefe commented:
> @tstuefe, @TobiHartmann, `InitialCodeCacheSize` is useless when static huge pages are used. When static huge pages are reserved, they are committed: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/memory/virtualspace.cpp#L248
> 
> ```
>   if (use_explicit_large_pages(page_size)) {
>     // System can't commit large pages i.e. use transparent huge pages and
>     // the caller requested large pages. To satisfy this request we use
>     // explicit large pages and these have to be committed up front to ensure
>     // no reservations are lost.
>     do {
> ```

Yes. For that case, the "lets at least do 8 pages so that we can commit as we go" logic makes no sense, and we may just as well live with one page, e.g. if pagesize = 1GB. I think that was what @TobiHartmann alluded to.

There is still the question of THPs. There, this logic still makes theoretical sense since THPs can be committed and uncommitted. But THPs (for now) only come in 2MB size. And as we established earlier, said logic contradicts what the java heap does.

My vote goes for just removing that logic.

eastig commented:
Hi @tstuefe, @TobiHartmann,
I'd like to discuss the changes I've made in `CodeCache::initialize` to support different pages sizes.
A problem with large pages is that the final page size is unknown till a successful reservation. A page size before a successful reservation is a preferred page size. 
Based on this I changed `CodeCache::initialize`:
1. Calculate a preferred page size. By default, use the one which allows magical 8 pages at least. If large pages are requested, the largest  page is used only if a user has specified `ReservedCodeCacheSize` in cmd. Most users use large pages for the Java heap not for CodeCache. So for large pages more than 2M, e.g. 1G, users will need to specify  `ReservedCodeCacheSize` .
2. Try to enable the segmented CodeCache. I moved the code from `CompilerConfig`. I think this is CodeCache responsibility to be or not to be segmented.
3. Initialise code heaps sizes.
4. As the preferred page size calculated in Step 1 might not be suitable for code heaps, we need to recalculate it with respect of code heaps sizes.
5. We reserve memory.
6. We create code heaps.  As the special reserved  space for CodeCache is fully committed (true for explicit huge pages), the initial sizes of code heaps is set to reserved sizes.

eastig commented:
Other changes:
- `os::can_execute_large_page_memory` is removed.
- `CodeCache::page_size(bool aligned = true, size_t min_pages = 1)` is removed because of no uses outside of CodeCache.

eastig commented:
Testing plan:
- Add new tests.
- Run tier1 - tier4:
  - Linux x86_64 and aarch64: normal pages, explicit huge pages and transparent huge pages.
  - Windows x86_64: normal pages and large pages.
  - MacOS x86_64 and aarch64: normal pages

tstuefe commented:
Sorry, no time to look at this in depth atm. This is suddenly a large patch, not sure why. I'd suggest to keep refactoring for a different RFE.

eastig commented:
> Sorry, no time to look at this in depth atm. This is suddenly a large patch, not sure why. I'd suggest to keep refactoring for a different RFE.

@tstuefe No worries. I'll do refactoring in different RFEs.
I made a fix which allows 1g pages for non-segmented CodeCache and 2M pages for segmented CodeCache.
Anyway 1g pages are not practical for segmented CodeCache because the size of CodeCache is limited to 2g. Also having 1g non-nmethod code heap makes no sense. I'll add a test tomorrow.

```
$ cat /proc/meminfo
...
HugePages_Total:     200
HugePages_Free:      200
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:        10895360 kB
DirectMap4k:      778136 kB
DirectMap2M:    33562624 kB
DirectMap1G:    363855872 kB

$ ./java -XX:+PrintCodeCache -XX:+UseLargePages -Xlog:pagesize=debug -Xmx1g -Xms1g -XX:+UseLargePages -XX:LargePageSizeInBytes=1g -version
[0.001s][info][pagesize] Static hugepage support:
[0.001s][info][pagesize]   hugepage size: 2M
[0.001s][info][pagesize]   hugepage size: 1G
[0.001s][info][pagesize]   default hugepage size: 2M
[0.001s][info][pagesize] Transparent hugepage (THP) support:
[0.001s][info][pagesize]   THP mode: madvise
[0.001s][info][pagesize]   THP pagesize: 2M
[0.001s][info][pagesize] Overriding default large page size (2M) using LargePageSizeInBytes: 1G
[0.001s][info][pagesize] UseLargePages=1, UseTransparentHugePages=0
[0.001s][info][pagesize] Large page support enabled. Usable page sizes: 4k, 2M, 1G. Default large page size: 1G.
[0.002s][debug][pagesize] Commit special mapping: 0x00007ff03ca00000, size=240M, page size=2M
[0.002s][info ][pagesize] CodeHeap 'non-nmethods':  min=4M max=10M base=0x00007ff043c00000 size=10M page_size=2M
[0.003s][info ][pagesize] CodeHeap 'profiled nmethods':  min=4M max=114M base=0x00007ff03ca00000 size=114M page_size=2M
[0.005s][info ][pagesize] CodeHeap 'non-profiled nmethods':  min=4M max=116M base=0x00007ff044600000 size=116M page_size=2M
[0.064s][debug][pagesize] Commit special mapping: 0x00000000c0000000, size=1G, page size=1G
[0.064s][info ][pagesize] Heap:  min=1G max=1G base=0x00000000c0000000 size=1G page_size=1G
[0.064s][debug][pagesize] Commit special mapping: 0x00007ff03a600000, size=2M, page size=2M
[0.064s][info ][pagesize] Block Offset Table: req_size=2M req_page_size=2M base=0x00007ff03a600000 size=2M page_size=2M
[0.064s][debug][pagesize] Commit special mapping: 0x00007ff03a200000, size=2M, page size=2M
[0.064s][info ][pagesize] Card Table: req_size=2M req_page_size=2M base=0x00007ff03a200000 size=2M page_size=2M
[0.064s][debug][pagesize] Commit special mapping: 0x00007ff039000000, size=16M, page size=2M
[0.064s][info ][pagesize] Mark Bitmap: req_size=16M req_page_size=2M base=0x00007ff039000000 size=16M page_size=2M
openjdk version "22-internal" 2024-03-19
OpenJDK Runtime Environment (fastdebug build 22-internal-adhoc.eastig.jdk)
OpenJDK 64-Bit Server VM (fastdebug build 22-internal-adhoc.eastig.jdk, mixed mode, sharing)
CodeHeap 'non-profiled nmethods': size=118784Kb used=16Kb max_used=16Kb free=118767Kb
 bounds [0x00007ff044600000, 0x00007ff044a00000, 0x00007ff04ba00000]
CodeHeap 'profiled nmethods': size=116736Kb used=120Kb max_used=120Kb free=116615Kb
 bounds [0x00007ff03ca00000, 0x00007ff03ce00000, 0x00007ff043c00000]
CodeHeap 'non-nmethods': size=10240Kb used=3690Kb max_used=3701Kb free=6549Kb
 bounds [0x00007ff043c00000, 0x00007ff044000000, 0x00007ff044600000]
 total_blobs=752 nmethods=72 adapters=591
 compilation: enabled
              stopped_count=0, restarted_count=0
 full_count=0

$ ./java -XX:+PrintCodeCache -XX:+UseLargePages -Xlog:pagesize=debug -Xmx1g -Xms1g -XX:+UseLargePages -XX:ReservedCodeCacheSize=1G -XX:InitialCodeCacheSize=1G -XX:LargePageSizeInBytes=1g -version
[0.001s][info][pagesize] Static hugepage support:
[0.001s][info][pagesize]   hugepage size: 2M
[0.001s][info][pagesize]   hugepage size: 1G
[0.001s][info][pagesize]   default hugepage size: 2M
[0.001s][info][pagesize] Transparent hugepage (THP) support:
[0.001s][info][pagesize]   THP mode: madvise
[0.001s][info][pagesize]   THP pagesize: 2M
[0.001s][info][pagesize] Overriding default large page size (2M) using LargePageSizeInBytes: 1G
[0.001s][info][pagesize] UseLargePages=1, UseTransparentHugePages=0
[0.001s][info][pagesize] Large page support enabled. Usable page sizes: 4k, 2M, 1G. Default large page size: 1G.
[0.002s][debug][pagesize] Commit special mapping: 0x00007f3680000000, size=1G, page size=1G
[0.002s][info ][pagesize] CodeCache:  min=1G max=1G base=0x00007f3680000000 size=1G page_size=1G
[0.002s][debug][pagesize] Commit special mapping: 0x00007f370b600000, size=8M, page size=2M
[0.435s][debug][pagesize] Commit special mapping: 0x00000000c0000000, size=1G, page size=1G
[0.435s][info ][pagesize] Heap:  min=1G max=1G base=0x00000000c0000000 size=1G page_size=1G
[0.435s][debug][pagesize] Commit special mapping: 0x00007f36fae00000, size=2M, page size=2M
[0.435s][info ][pagesize] Block Offset Table: req_size=2M req_page_size=2M base=0x00007f36fae00000 size=2M page_size=2M
[0.435s][debug][pagesize] Commit special mapping: 0x00007f36faa00000, size=2M, page size=2M
[0.435s][info ][pagesize] Card Table: req_size=2M req_page_size=2M base=0x00007f36faa00000 size=2M page_size=2M
[0.435s][debug][pagesize] Commit special mapping: 0x00007f36f9800000, size=16M, page size=2M
[0.435s][info ][pagesize] Mark Bitmap: req_size=16M req_page_size=2M base=0x00007f36f9800000 size=16M page_size=2M
openjdk version "22-internal" 2024-03-19
OpenJDK Runtime Environment (fastdebug build 22-internal-adhoc.eastig.jdk)
OpenJDK 64-Bit Server VM (fastdebug build 22-internal-adhoc.eastig.jdk, mixed mode, sharing)
CodeCache: size=1048576Kb used=3884Kb max_used=3884Kb free=1044691Kb
 bounds [0x00007f3680000000, 0x00007f36c0000000, 0x00007f36c0000000]
 total_blobs=752 nmethods=72 adapters=591
 compilation: enabled
              stopped_count=0, restarted_count=0
 full_count=0
```

eastig commented:
Linux x86_64, 200 2MB pages, 10 1GB pages, the following tests passed:
```
hotspot/jtreg/compiler/codecache/jmx/InitialAndMaxUsageTest.java
hotspot/jtreg/gc/arguments/TestAlignmentToUseLargePages.java
hotspot/jtreg/gc/arguments/TestNewRatioFlag.java
hotspot/jtreg/gc/arguments/TestNewSizeFlags.java
hotspot/jtreg/gc/g1/numa/TestG1NUMATouchRegions.java
hotspot/jtreg/gc/g1/TestLargePageUseForAuxMemory.java
hotspot/jtreg/gc/g1/TestLargePageUseForHeap.java
hotspot/jtreg/gc/metaspace/TestMetaspaceSizeFlags.java
hotspot/jtreg/gc/shenandoah/options/TestLargePages.java
hotspot/jtreg/gc/shenandoah/options/TestLargePagesWithSmallHeap.java
hotspot/jtreg/gc/shenandoah/TestHeapUncommit.java
hotspot/jtreg/gc/TestAllocateHeapAtMultiple.java
hotspot/jtreg/gc/TestNUMAPageSize.java
hotspot/jtreg/gc/TestSystemGC.java
hotspot/jtreg/gtest/LargePageGtests.java
hotspot/jtreg/runtime/cds/appcds/CommandLineFlagCombo.java
hotspot/jtreg/runtime/cds/appcds/SharedRegionAlignmentTest.java
hotspot/jtreg/runtime/cds/appcds/sharedStrings/InternSharedString.java
hotspot/jtreg/runtime/cds/appcds/sharedStrings/LargePages.java
hotspot/jtreg/runtime/CompressedOops/CompressedClassPointers.java
hotspot/jtreg/runtime/CompressedOops/UseCompressedOops.java
hotspot/jtreg/runtime/memory/LargePages/TestLargePagesFlags.java
hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java
hotspot/jtreg/runtime/os/TestTracePageSizes.java
hotspot/jtreg/vmTestbase/gc/memory/LargePagesTest/LargePagesTest.java
```

The test `hotspot/jtreg/compiler/codecache/CheckLargePages.java` passed with [the regex fix](https://github.com/yftsai/jdk/compare/master...yftsai:jdk:JDK-8317831#diff-9894fc5c648bf7884ce182cd1d162ebac09789d20bc2ef6614bb76b40e3f121dR65).

eastig commented:
@tstuefe, @TobiHartmann, could you have a look?

TobiHartmann reviewed (approved):
Looks reasonable to me but I'm not an expert in this code (anymore). @tstuefe should have a look as well.

tstuefe commented on test/hotspot/jtreg/compiler/codecache/CheckLargePages.java:
```diff
@@ -44,35 +44,83 @@
+                "-version");
+        OutputAnalyzer out = new OutputAnalyzer(pb.start());
+        out.shouldMatch("Code cache size too small for \\S* pages\\. Reverting to smaller page size \\((\\S*)\\)\\.");
+        out.shouldHaveExitValue(0);
```
I don't get this.

The new version (which I like, btw) now assumes min_pages=1 if InitialCCS == ReservedCCS. Should this not then work? I would have expected to get a code cache with 2x1G pages. I would not have expected a "code cache size too small" warning.

eastig replied:
> The new version (which I like, btw) now assumes min_pages=1 if InitialCCS == ReservedCCS.

It is only for the non-segmented CodeCache. Here we explicitly specify to use the segmented CodeCache. As such, we need to split 2G among three code heaps. InitialCCS is tricky in such a case, with the segmented CodeCache, the initial size of a code heap is the min of InitialCCS and the code heap size. We want the 2G segmented CodeCache to have 8 pages at least which is not possible with 1G pages. The number 8 is hard-coded. 

I think calculating the minimum number of pages is not easy for the segmented CodeCache.
We have:
- InitialCodeCacheSize
- ReservedCodeCacheSize
- LargePageSizeInBytes
- NonNMethodCodeHeapSize
- ProfiledCodeHeapSize
- NonProfiledCodeHeapSize

It's too many parameters for users.

A possible configuration: one 1G page for the non-nmethod code heap and the profiled code heap, another 1G page for the non-profiled code heap.

BTW, this is the original first test.

tstuefe replied:
Okay, I see.

We should probably allow the segmented code cache, in its entirety, to live on one or a few super large pages if we desire to; but I guess that would be too much changes for now. Not for this RFE.

tstuefe reviewed (commented):
The code itself is ok and a nice simplification. I have questions about the test, see below.

tstuefe reviewed (approved):
Looks good!

eastig commented:
/integrate
